### PR TITLE
Upgrade to ethereumjs-vm 2.0.1

### PR DESCRIPTION
**Pull request by simondlr**
See: #197.

### CODE REVIEW

**Comment by redsquirrel** [comment]
This resolves a problem that has been plaguing a project that @drcode have been working on. It manifested itself as an "out of gas" error. I just confirmed that this PR resolves our issue! 🎉

**Comment by tcoulter** [comment]
Is this still in progress? (it's tagged as "in progress"). Will pull and run the tests.

**Comment by simondlr** [comment]
Oh. Sorry. Thought "in progress" meant, as in, reviewing "in progress".

**Comment by tcoulter** [comment]
:+1: I don't know what it means. :) I thought it meant you're still working on it.

Either way, upon first look it appears ethereumjs-vm might have an issue when pulling the blockhash within solidity code. Need to take a deep dive before this can be merged.

**Comment by simondlr** [comment]
> I thought it meant you're still working on it.

Oh, it was added automatically. Might be github doing this. :D

> Either way, upon first look it appears ethereumjs-vm might have an issue when pulling the blockhash within solidity code.

Yeah, might be related to some other changes.

**Comment by tcoulter** [comment]
@johnmcdowall I've assigned this one to you. I've created a new branch called `simondlr-master` that includes Simon's work and a couple test fixes. 

There are still some tests that fail which need to be looked at before we can merge this branch. From what I can tell, the tests that fail are all related to processing contracts that asks for a block's hash from within the contract code itself. Somehow the block hash returned in code doesn't match what we expect it to be. My hunch is this is a problem within the newest ethereumjs-vm itself; if so, it should lead you to the `BLOCKHASH` opcode here: https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/opFns.js#L326. When I've previously debugged this, I've used `npm link` on my own version of ethereumjs-vm to better understand the code the `BLOCKHASH` opcode is calling (and hence why the wrong hash is returned). That said, this could totally be an issue with the TestRPC as well.

The main work here is a bunch of debugging. Getting a handle on all the moving parts is likely the hardest. Let me know if you have any questions.

**Comment by tcoulter** [comment]
@johnmcdowall PS: [This](https://github.com/ethereumjs/testrpc/blob/master/test/requests.js#L392) is the test you should fix first. There's an identical test that performs the same actions but within the context of chain forking; ignore that one. Fixing the non-forked version of the test should likely fix the forked one as well.

**Comment by jmdfm** [comment]
Right, so the new interface as well as changing to opts, made passing a blockchain an option where before it was required. This is why the 3 tests are failing, we're not passing in the blockchain double as the `blockchain` opt, so the VM is creating one every test run which results in block 0 having a different hash every time.

**Comment by jmdfm** [comment]
See PR #205

File: lib/blockchain_double.js [hunk 1]
```diff
@@ -21,7 +21,8 @@ function BlockchainDouble(options) {
 
   this.stateTrie = options.trie || new Trie();
 
-  this.vm = options.vm || new VM(this.stateTrie, this, {
+  this.vm = options.vm || new VM({
+    state: this.stateTrie,
```
Comment (axic) [inline lib/blockchain_double.js:25]: The `this` parameter was missed, it should be here as `blockchain: this,`.

Comment (jmdfm) [reply lib/blockchain_double.js:25]: Haha you just jinxed me on PR #205 :)

**Comment by jmdfm** [comment]
Closing this in favour of #205 which is ready

**Pull request closed without merging**